asmcomp/cmmgen.ml

```diff
@@ -1982,6 +1982,17 @@ and transl_prim_1 env p arg dbg =
                    [untag_int (transl env arg) dbg],
                    dbg))
               dbg
+  | Pcoerce coerce_to ->
```

Comment:
I don't understand why `Coerce_to_int` is needed; can you explain?

I was thinking perhaps we could have `Pensure_const_ptr` or something.  However then I wondered about the following.  At present we only force the register typing (using `Uconst_ptr`) in the case where the scrutinee of the match turns out to be constant.  So what about if the scrutinee is some variable which is known (either by its type or because it's defined by some integer-only expression, perhaps) to be always immediate but which is deemed inconstant (i.e. not manifestly `Uconst_...`)?  With this patch we need to ensure that such a variable ends up having a "val" not an "int" register type.  It is not at all obvious from the code at present whether this is always going to happen and I worry that it might not in some cases:

- Value kinds (one of which is "int") are propagated through Clambda nowadays on `Clet`; however Selectgen infers the register kind of the bound variable of a "let" from its defining expression.  This seems the most likely point for it to break, to me, since perhaps that defining expression could be some operation that `Selectgen` always believes has an integer result (which it does, but...).
- Function parameters in Selectgen are assigned register types based on what comes through from Cmmgen; I think these will always be "val" for OCaml functions at the moment, but perhaps that might change in the future.

One more robust solution might be to push the coercion further down (into Selectgen), but it might obstruct optimizations in Cmmgen then.

@chambart Do you think this reasoning is correct, and do you have any thoughts?

The rest of the patch is OK modulo this problem.  I believe this work will be covered by the Jane Street corporate CLA.